### PR TITLE
LUCENE-9319: Clean up package name conflicts for sandbox module

### PR DESCRIPTION
**Pull request by mocobeta**
"sandbox" module shares package names o.a.l.codecs, o.a.l.document, and o.a.l.search with lucene-core. We have two choices to clean up these conflicts:

- Move them under `o.a.l.sandbox` (i.e., `o.a.l.sandbox.codecs`, `o.a.l.sandbox.document`, and `o.a.l.sandbox.search`)  and relax visibility of some classes/methods in "core" to public/protected to allow classes in sandbox module access those (currently package-private) classes/methods.
- Move them to core module to keep package-private visibility in core module as is.

Here, I took the first option for all classes for the first step to resolve the conflicts. But that may be a rather wild approach; some classes probably should be moved to core (option 2).

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/document/RangeFieldQuery.java [hunk 1]
```diff
@@ -41,7 +41,7 @@
 /**
  * Query class for searching {@code RangeField} types by a defined {@link Relation}.
  */
-abstract class RangeFieldQuery extends Query {
+public abstract class RangeFieldQuery extends Query {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/document/RangeFieldQuery.java:44]: This is used in o.a.l.sandbox.document.LatLonBoundingBox.

File: lucene/core/src/java/org/apache/lucene/document/RangeFieldQuery.java [hunk 2]
```diff
@@ -57,7 +57,7 @@
   final int bytesPerDim;
 
   /** Used by {@code RangeFieldQuery} to check how each internal or leaf node relates to the query. */
-  enum QueryType {
+  public enum QueryType {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/document/RangeFieldQuery.java:60]: This is used in o.a.l.sandbox.document.LatLonBoundingBox.

File: lucene/core/src/java/org/apache/lucene/document/RangeFieldQuery.java [hunk 3]
```diff
@@ -228,7 +228,7 @@ boolean matches(byte[] queryPackedValue, byte[] packedValue, int numDims, int by
    * @param ranges encoded range values; this is done by the {@code RangeField} implementation
    * @param queryType the query relation
    */
-  RangeFieldQuery(String field, final byte[] ranges, final int numDims, final QueryType queryType) {
+  protected RangeFieldQuery(String field, final byte[] ranges, final int numDims, final QueryType queryType) {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/document/RangeFieldQuery.java:231]: This is used in o.a.l.sandbox.document.LatLonBoundingBox.

File: lucene/core/src/java/org/apache/lucene/search/ExactPhraseMatcher.java [hunk 4]
```diff
@@ -33,7 +33,8 @@
 import org.apache.lucene.search.similarities.Similarity.SimScorer;
 import org.apache.lucene.util.PriorityQueue;
 
-final class ExactPhraseMatcher extends PhraseMatcher {
+/** Expert: Find exact phrases */
+public final class ExactPhraseMatcher extends PhraseMatcher {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/ExactPhraseMatcher.java:37]: This is used in o.a.l.sandbox.search.PhraseWildcardQuery.

File: lucene/core/src/java/org/apache/lucene/search/ExactPhraseMatcher.java [hunk 5]
```diff
@@ -50,7 +51,8 @@ public PostingsAndPosition(PostingsEnum postings, int offset) {
   private final DocIdSetIterator approximation;
   private final ImpactsDISI impactsApproximation;
 
-  ExactPhraseMatcher(PhraseQuery.PostingsAndFreq[] postings, ScoreMode scoreMode, SimScorer scorer, float matchCost) {
+  /** Expert: Creates ExactPhraseMatcher instance */
+  public ExactPhraseMatcher(PhraseQuery.PostingsAndFreq[] postings, ScoreMode scoreMode, SimScorer scorer, float matchCost) {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/ExactPhraseMatcher.java:55]: Same avobe.

File: lucene/core/src/java/org/apache/lucene/search/HitQueue.java [hunk 6]
```diff
@@ -19,7 +19,8 @@
 
 import org.apache.lucene.util.PriorityQueue;
 
-final class HitQueue extends PriorityQueue<ScoreDoc> {
+/** Expert: Priority queue containing hit docs */
+public final class HitQueue extends PriorityQueue<ScoreDoc> {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/HitQueue.java:23]: This is used in o.a.l.sandbox.search.LargeNumHitsTopDocsCollector.

File: lucene/core/src/java/org/apache/lucene/search/HitQueue.java [hunk 7]
```diff
@@ -59,7 +60,7 @@
    * @param prePopulate
    *          specifies whether to pre-populate the queue with sentinel values.
    */
-  HitQueue(int size, boolean prePopulate) {
+  public HitQueue(int size, boolean prePopulate) {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/HitQueue.java:63]: Same above.

File: lucene/core/src/java/org/apache/lucene/search/MultiPhraseQuery.java [hunk 8]
```diff
@@ -410,7 +410,7 @@ private boolean termArraysEquals(Term[][] termArrays1, Term[][] termArrays2) {
    * <p>
    * Note: positions are merged during freq()
    */
-  static class UnionPostingsEnum extends PostingsEnum {
+  public static class UnionPostingsEnum extends PostingsEnum {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/MultiPhraseQuery.java:413]: This is used in o.a.l.sandbox.search.PhraseWildcardQuery.

File: lucene/core/src/java/org/apache/lucene/search/MultiPhraseQuery.java [hunk 9]
```diff
@@ -423,7 +423,7 @@ private boolean termArraysEquals(Term[][] termArrays1, Term[][] termArrays2) {
     /** list of subs (unordered) */
     final PostingsEnum[] subs;
 
-    UnionPostingsEnum(Collection<PostingsEnum> subs) {
+    public UnionPostingsEnum(Collection<PostingsEnum> subs) {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/MultiPhraseQuery.java:426]: Same avobe.

File: lucene/core/src/java/org/apache/lucene/search/MultiPhraseQuery.java [hunk 10]
```diff
@@ -575,17 +575,17 @@ private void growArray() {
     }
   }
 
-  // Slower version of UnionPostingsEnum that delegates offsets and positions, for
-  // use by MatchesIterator
-  static class UnionFullPostingsEnum extends UnionPostingsEnum {
+  /** Slower version of UnionPostingsEnum that delegates offsets and positions, for
+   use by MatchesIterator */
+  public static class UnionFullPostingsEnum extends UnionPostingsEnum {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/MultiPhraseQuery.java:580]: This is used in o.a.l.sandbox.search.PhraseWildcardQuery.

File: lucene/core/src/java/org/apache/lucene/search/MultiPhraseQuery.java [hunk 11]
```diff
@@ -575,17 +575,17 @@ private void growArray() {
...
 
     int freq = -1;
     boolean started = false;
 
     final PriorityQueue<PostingsAndPosition> posQueue;
     final Collection<PostingsAndPosition> subs;
 
-    UnionFullPostingsEnum(List<PostingsEnum> subs) {
+    public UnionFullPostingsEnum(List<PostingsEnum> subs) {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/MultiPhraseQuery.java:588]: Same above.

File: lucene/core/src/java/org/apache/lucene/search/Multiset.java [hunk 12]
```diff
@@ -29,13 +29,13 @@
  * Iteration order is not specified.
  * @lucene.internal
  */
-final class Multiset<T> extends AbstractCollection<T> {
+public final class Multiset<T> extends AbstractCollection<T> {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/Multiset.java:32]: This is used in o.a.l.sandbox.search.CoveringQuery.

File: lucene/core/src/java/org/apache/lucene/search/Multiset.java [hunk 13]
```diff
@@ -29,13 +29,13 @@
...
 
   private final Map<T, Integer> map = new HashMap<>();
   private int size;
 
   /** Create an empty {@link Multiset}. */
-  Multiset() {
+  public Multiset() {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/Multiset.java:38]: Same above.

File: lucene/core/src/java/org/apache/lucene/search/PhraseMatcher.java [hunk 14]
```diff
@@ -26,7 +26,7 @@
  * relevant document, then call {@link #reset()}.  Clients can then call
  * {@link #nextMatch()} to iterate over the matches
  */
-abstract class PhraseMatcher {
+public abstract class PhraseMatcher {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/PhraseMatcher.java:29]: This is used in o.a.l.sandbox.search.PhraseWildcardQuery.

File: lucene/core/src/java/org/apache/lucene/search/PhraseQuery.java [hunk 15]
```diff
@@ -304,13 +304,15 @@ public void visit(QueryVisitor visitor) {
     v.consumeTerms(this, terms);
   }
 
-  static class PostingsAndFreq implements Comparable<PostingsAndFreq> {
+  /** Term postings and position information for phrase matching */
+  public static class PostingsAndFreq implements Comparable<PostingsAndFreq> {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/PhraseQuery.java:308]: This is used in o.a.l.sandbox.search.PhraseWildcardQuery.

File: lucene/core/src/java/org/apache/lucene/search/PhraseQuery.java [hunk 16]
```diff
@@ -413,7 +415,7 @@ public boolean equals(Object obj) {
    *  This is for use by {@link TwoPhaseIterator#matchCost} implementations.
    *  @param termsEnum The term is the term at which this TermsEnum is positioned.
    */
-  static float termPositionsCost(TermsEnum termsEnum) throws IOException {
+  public static float termPositionsCost(TermsEnum termsEnum) throws IOException {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/PhraseQuery.java:418]: This is used in o.a.l.sandbox.search.PhraseWildcardQuery.

File: lucene/core/src/java/org/apache/lucene/search/PhraseWeight.java [hunk 17]
```diff
@@ -23,13 +23,15 @@
 import org.apache.lucene.search.similarities.Similarity;
 import org.apache.lucene.search.similarities.Similarity.SimScorer;
 
-abstract class PhraseWeight extends Weight {
+/** Expert: Weight class for phrase matching */
+public abstract class PhraseWeight extends Weight {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/PhraseWeight.java:27]: This is used in o.a.l.sandbox.search.PhraseWildcardQuery.

File: lucene/core/src/java/org/apache/lucene/search/SloppyPhraseMatcher.java [hunk 18]
```diff
@@ -53,7 +53,7 @@
  * would get same score as "g f"~2, although "c b"~2 could be matched twice.
  * We may want to fix this in the future (currently not, for performance reasons).
  */
-final class SloppyPhraseMatcher extends PhraseMatcher {
+public final class SloppyPhraseMatcher extends PhraseMatcher {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/SloppyPhraseMatcher.java:56]: This is used in o.a.l.sandbox.search.PhraseWildcardQuery.

File: lucene/core/src/java/org/apache/lucene/search/SloppyPhraseMatcher.java [hunk 19]
```diff
@@ -81,7 +81,7 @@
   private boolean positioned;
   private int matchLength;
 
-  SloppyPhraseMatcher(PhraseQuery.PostingsAndFreq[] postings, int slop, ScoreMode scoreMode, SimScorer scorer, float matchCost, boolean captureLeadMatch) {
+  public SloppyPhraseMatcher(PhraseQuery.PostingsAndFreq[] postings, int slop, ScoreMode scoreMode, SimScorer scorer, float matchCost, boolean captureLeadMatch) {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/SloppyPhraseMatcher.java:84]: Same above.

File: lucene/core/src/java/org/apache/lucene/search/TermScorer.java [hunk 20]
```diff
@@ -25,7 +25,7 @@
 
 /** Expert: A <code>Scorer</code> for documents matching a <code>Term</code>.
  */
-final class TermScorer extends Scorer {
+public final class TermScorer extends Scorer {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/TermScorer.java:28]: This is used in o.a.l.sandbox.search.BM25FQuery.

File: lucene/core/src/java/org/apache/lucene/search/TermScorer.java [hunk 21]
```diff
@@ -35,7 +35,7 @@
   /**
    * Construct a {@link TermScorer} that will iterate all documents.
    */
-  TermScorer(Weight weight, PostingsEnum postingsEnum, LeafSimScorer docScorer) {
+  public TermScorer(Weight weight, PostingsEnum postingsEnum, LeafSimScorer docScorer) {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/TermScorer.java:38]: Same above.

File: lucene/core/src/java/org/apache/lucene/search/TermScorer.java [hunk 22]
```diff
@@ -60,7 +60,8 @@ public int docID() {
     return postingsEnum.docID();
   }
 
-  final int freq() throws IOException {
+  /** Returns term frequency in the current document. */
+  public final int freq() throws IOException {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/TermScorer.java:64]: Same above.

File: lucene/core/src/java/org/apache/lucene/search/TopDocsCollector.java [hunk 23]
```diff
@@ -35,7 +35,7 @@
 
   /** This is used in case topDocs() is called with illegal parameters, or there
    *  simply aren't (enough) results. */
-  protected static final TopDocs EMPTY_TOPDOCS = new TopDocs(new TotalHits(0, TotalHits.Relation.EQUAL_TO), new ScoreDoc[0]);
+  public static final TopDocs EMPTY_TOPDOCS = new TopDocs(new TotalHits(0, TotalHits.Relation.EQUAL_TO), new ScoreDoc[0]);
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/TopDocsCollector.java:38]: This is used in o.a.l.sandbox.search.LargeNumHitsTopDocsCollector.

File: lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java [hunk 24]
```diff
@@ -38,9 +38,10 @@
  */
 public abstract class TopScoreDocCollector extends TopDocsCollector<ScoreDoc> {
 
-  abstract static class ScorerLeafCollector implements LeafCollector {
+  /** Scorable leaf collector */
+  public abstract static class ScorerLeafCollector implements LeafCollector {
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java:42]: This is used in o.a.l.sandbox.search.LargeNumHitsTopDocsCollector.

File: lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java [hunk 25]
```diff
@@ -38,9 +38,10 @@
...
 
-    Scorable scorer;
+    protected Scorable scorer;
```
Comment (mocobeta) [inline lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java:44]: Same above.

**Review by dweiss — APPROVED** [review]
It does suck a bit that some of these classes need to be made public. But then - if the sandbox is using it then perhaps they should be publicly accessible for others to experiment with. 

Ideally, these modules should be open to one another but closed to the public. Until we switch to the module system this won't be possible. 

I say let's clean up package splits, even if at the cost of making some things visible. +1 from me.

**Comment by mocobeta** [comment]
@dweiss thanks for taking a look at this.

@nknize  @bruno-roustant @atris  @jpountz  @jimczi  
I'm trying to clean up shared packages between sandbox and lucene-core. But for now I have little knowledge about those classes that were moved in this pull request; I think you're the ones of first authors of the experimental classes, I'd like to ask your thoughts.
If you think some classes should be moved to lucene-core (instead of opening the classes/methods in core module to public), or have any concerns about changes I made here, would you leave comments?

I'll wait one more week, then if there is no objection, merge it into master.

**Review by bruno-roustant — APPROVED** [review]
I mainly reviewed PhraseWildcardQuery. I agree with the choice of opening the visibility of package private classes/methods in o.a.l.search. That said I don't know precisely the context of keeping those classes/methods package private until now. I suppose it is the principle of least visibility, but now that we see we need them public it is ok.

**Review by jpountz — COMMENTED** [review]
Maybe we should look into cleaning up this sandbox as well.

For instance LatLonBoundingBox predates LatLonShape and I'm not sure it still serves a purpose. Sure it's more efficient for someone who only wishes to index bounding boxes, but I'm not sure whether this is a frequent enough use-case to be worth maintaining. What do you think @iverase @nknize @dsmiley ?

For classes in core that we're making public only to support some sandbox classes, can you add the `@lucene.internal` tag to convey that there are no backward compatibility guarantees for these classes?

**Comment by mocobeta** [comment]
I have added `@lucene.internal` to classes in core that are made publicly accessible in the changes.

Should we refactor or retire some classes in sandbox? Maybe we can open this PR for a while to discuss which APIs in core will be publicly accessible for future experimentation. (`misc` module has the same challenge too...)

**Comment by mocobeta** [comment]
> Maybe we can keep open this PR for a while to discuss which APIs in core will be publicly accessible for future experimentation.

Ok, I don't think we can reach the conclusion about that now. :) Once we make the shift to Java module system, then we can reconsider which APIs should be exported to 3rd party modules. In that sense, I think changes here are temporal measures until then and `@lucene.internal` tag will help developers to notice that the classes with it will not be guaranteed backward compatibility in the future releases.
This got approvals, so I'm going to merge it to the master if there is no objection/comment in the next few days. Let me know if there are classes in sandbox that should be retired at 9.0 release.

**Comment by mocobeta** [comment]
Just curious... does it make sense we have BM25FQuery in core? I thought it is a missing peace for Lucene as a full-text search engine with multi fields, but I have not yet closely looked at the current implementation and of course it can stay in sandbox until it's matured to moved.

**Pull request merged**